### PR TITLE
[10.0][IMP] queue_job: Don't prefetch fields when vacuum + bumpversion

### PR DESCRIPTION
**Pull request by rousseldenis**
@guewen @lmignon 

Did this as sometimes, the amount of jobs is huge and there is no need fetching fields.

### CODE REVIEW

**Comment by rousseldenis** [comment]
@guewen I bumped version as it was not done for long. But in separate commit if you disagree.

File: queue_job/models/queue_job.py [hunk 1]
```diff
@@ -226,7 +226,7 @@ def autovacuum(self):
         Called from a cron.
         """
         deadline = datetime.now() - timedelta(days=self._removal_interval)
-        jobs = self.search(
+        jobs = self.with_context(prefetch_fields=False).search(
```
Comment (guewen) [inline queue_job/models/queue_job.py]: I was sure that the prefetch only happen the first time you access a field (other than id)?

Comment (rousseldenis) [reply queue_job/models/queue_job.py]: I will try to objective that with a test script (not in the module)

**Comment by rousseldenis** [comment]
@guewen I think you're right.
The problem is the timeout when your backlog is huge. I was inspired by odoo orderpoint runner (https://github.com/odoo/odoo/blob/10.0/addons/stock/models/procurement.py#L308) to make sets of jobs to unlink and commit.
I'm not sure of all the parts I've done, so if you have suggestions, I will implement them !

File: queue_job/models/queue_job.py [hunk 2]
```diff
@@ -220,16 +220,39 @@ def _needaction_domain_get(self):
...
+    def autovacuum(self, jobs_set=1000, use_new_cursor=False):
         """ Delete all jobs done since more than ``_removal_interval`` days.
-
+        Uses sets of jobs to ensure that no timeout occurs
         Called from a cron.
         """
+        job_obj = self.env['queue.job']
         deadline = datetime.now() - timedelta(days=self._removal_interval)
-        jobs = self.search(
+        jobs_ids = self.with_context(prefetch_fields=True).search(
```
Comment (lmignon) [inline queue_job/models/queue_job.py]: @rousseldenis IMO prefetch_field is useless

File: queue_job/models/queue_job.py [hunk 3]
```diff
@@ -220,16 +220,39 @@ def _needaction_domain_get(self):
...
+            jobs_ids = jobs_ids[jobs_set:]
+
+            jobs.unlink()
+
+            try:
+                if use_new_cursor:
+                    cr.commit()
+            except OperationalError:
+                if use_new_cursor:
+                    cr.rollback()
```
Comment (lmignon) [inline queue_job/models/queue_job.py:247]: @rousseldenis IMO all the logic should be enclosed into a try except finally block
```python
    if use_new_cursor:
        cr = registry(self._cr.dbname).cursor()
    try:
        self = self.with_env(self.env(cr=cr))
        jobs = job_obj.browse(jobs_ids[:jobs_set])
        jobs_ids = jobs_ids[jobs_set:]
        jobs.unlink()
        if use_new_cursor:
            cr.commit()
    except OperationalError:
         if use_new_cursor:
            cr.rollback()
            raise
    finally:
        if use_new_cursor:
            cr.close()
```

**Review by lmignon — CHANGES_REQUESTED** [review]

**Comment by rousseldenis** [comment]
@lmignon @guewen

**Review by lmignon — COMMENTED** [review]
@rousseldenis Can you modify the changlog PLZ https://github.com/OCA/queue/blob/10.0/queue_job/README.rst#changelog

**Comment by rousseldenis** [comment]
@guewen As suggested by @lmignon, should I fill in changelog as it should includes all commits descriptions since the previous version?

P.S.: The modification has been tested on a real database.

**Comment by sbidoul** [comment]
@rousseldenis are you sure the benefit is worth the additional code complexity?
I understood you faced the issue in a real pathological situation (which could be resolved manually by a sql delete). When the vacuum cron runs regularly nobody faces the issue, even on the largest queue job delployments AFAIK.

**Comment by rousseldenis** [comment]
@sbidoul You have to handle also mail messages. So, in fact there should be multiple sql deletes.

**Pull request closed without merging**